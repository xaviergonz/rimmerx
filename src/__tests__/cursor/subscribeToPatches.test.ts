import { PatchListener, rollbackUpdate, subscribeToPatches, update } from "../..";
import { $data, $firstUser } from "./testbed";

test("subscribeToPatches", () => {
  let calls = 0;
  const listener: PatchListener = (patches, inversePatches) => {
    calls++;
    expect({ patches, inversePatches }).toMatchSnapshot();
  };

  expect(() => subscribeToPatches($firstUser, listener)).toThrow("patch subscription can only be done on root cursors");

  const disposer = subscribeToPatches($data, listener);

  expect(calls).toBe(0);

  update($firstUser, u => {
    u.name = "james";
    throw rollbackUpdate;
  });
  expect(calls).toBe(0);

  update($firstUser, u => {
    u.name = "john";
  });
  expect(calls).toBe(1);

  disposer();
  update($firstUser, u => {
    u.name = "james";
  });
  expect(calls).toBe(1);
});
