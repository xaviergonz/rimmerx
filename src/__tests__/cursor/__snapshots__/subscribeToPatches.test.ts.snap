// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscribeToPatches 1`] = `
Object {
  "inversePatches": Array [
    Object {
      "op": "replace",
      "path": Array [
        "users",
        0,
        "name",
      ],
      "value": "first",
    },
  ],
  "patches": Array [
    Object {
      "op": "replace",
      "path": Array [
        "users",
        0,
        "name",
      ],
      "value": "john",
    },
  ],
}
`;
